R/dataframe-s3-base.R

```diff
@@ -121,3 +121,245 @@ head.polars_data_frame <- function(x, n = 6L, ...) x$head(n = n)
+        )
+      )
+    )
+  }
+
+  if (is.numeric(j) && !rlang::is_integerish(j)) {
+    abort(
+      c(
+        sprintf("Can't subset columns with `%s`.", deparse(j_arg)),
+        "x" = "Can't convert from `j` <double> to <integer> due to loss of precision."
+      )
+    )
+  }
+
+  # NA values are accepted for rows but not columns.
+  if (anyNA(j)) {
+    na_val <- which(is.na(j))
+    abort(
+      c(
+        sprintf("Can't subset columns with `%s`.", deparse(j_arg)),
+        "x" = sprintf("Subscript `%s` can't contain missing values.", deparse(j_arg)),
+        "x" = sprintf(
+          "It has missing value(s) at location %s.",
+          oxford_comma(na_val, final = "and")
+        )
+      )
+    )
+  }
+
+  if (!is_null(j)) {
+    # Can be:
+    # - numeric but cannot beyond the number of columns, and cannot mix positive
+    #   and negative indices
+    # - logical but must be of length 1 or number of columns
+    # - character
+    if (is.numeric(j)) {
+      wrong_locs <- j[j > length(cols)]
+      if (length(wrong_locs) > 0) {
+        abort(
+          c(
+            "Can't subset columns past the end.",
+            "i" = sprintf("Location(s) %s don't exist.", oxford_comma(wrong_locs, final = "and")),
+            "i" = sprintf("There are only %s columns.", length(cols))
+          )
+        )
+      }
+      if (all(j < 0)) {
+        j <- setdiff(seq_len(length(cols)), abs(j))
+      } else if (any(j < 0) && any(j > 0)) {
+        sign_start <- sign(j[j != 0])[1]
+        loc <- if (sign_start == -1) {
+          which(sign(j) == 1)[1]
+        } else if (sign_start == 1) {
+          which(sign(j) == -1)[1]
+        }
+        abort(
+          c(
+            "!" = sprintf("Can't subset columns with `%s`.", deparse(j_arg)),
+            "x" = "Negative and positive locations can't be mixed.",
+            "i" = sprintf(
+              "Subscript `%s` has a %s value at location %s.",
+              deparse(j_arg),
+              if (sign_start == 1) "negative" else "positive",
+              loc
+            )
+          )
+        )
+      }
+      to_select <- cols[j]
+    } else if (is_character(j)) {
+      to_select <- j
+    } else if (is_logical(j)) {
+      if (length(j) == 1) {
+        to_select <- cols
+      } else if (length(j) == length(cols)) {
+        to_select <- cols[j]
+      } else {
+        abort(
+          c(
+            "!" = sprintf("Can't subset columns with `%s`.", deparse(j_arg)),
+            "i" = sprintf(
+              "Logical subscript `%s` must be size 1 or %s, not %s",
+              deparse(j_arg),
+              length(cols),
+              length(j)
+            )
+          )
+        )
+      }
+    }
+
+    # Don't do this too early because x[, TRUE, drop = TRUE] mustn't drop if
+    # x has more than 1 column
+    if (length(to_select) > 1) {
+      drop <- FALSE
+    }
+
+    x <- x$select(to_select)
+  }
+
+  if (isTRUE(drop)) {
+    as.list(x)
+  } else {
```

Comment:
Is this correct?
I was thinking that if `drop=TRUE`, it should return Series if there is only one column, otherwise it should do nothing.

Reply:
Actually, when we handle columns just above, we set `drop <- FALSE` if there's more than one column selected. This is also tested with `test[, c("a", "b"), drop = TRUE]` (but it's easy to miss that in the list of tests).

Reply:
Ah ok but we returned a list containing a single Series instead of just returning the Series, I fixed that.